Generalize noncontiguous tests to several outputs

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67833
* #68188
* __->__ #67996
* #68523
* #68522
* #69909
* #69908

This is necessary for most matrix decompositions in `linalg`.

cc @mruberry

Differential Revision: [D33774418](https://our.internmc.facebook.com/intern/diff/D33774418)

lezcano committed:
Generalize noncontiguous tests to several outputs

This is necessary for most matrix decompositions in `linalg`.

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -2376,27 +2376,36 @@ def noncontiguous_like(t):
+    if zero_dim:
+        t = t.unsqueeze(0)
+
+    # Handle conj bit. repeat_interleave resolves the conj, which is a bit annoying.
```
Nice catch

anjali411 replied:
Actually I don't see any reason why we need to resolve conjugation or negation for `repeat_interleave`. We should register the `repeat_interleave` as fallthrough for both `Conjugate` and `Negative` dispatch key and simply unset the conj/neg bit for `self` before the computation and then reset it.  `repeats` is guaranteed to be an int tensor so we don't have to worry about it.

anjali411 replied:
It might be worth making this change in this PR so that you don't have to add special conj/neg handling

mruberry commented on torch/testing/_internal/common_utils.py:
```diff
@@ -2376,27 +2376,36 @@ def noncontiguous_like(t):
+        strides[-1] *= 2
+        result.set_(result.storage(), result.storage_offset(), t.size(), stride=tuple(strides))
     result.requires_grad_(t.requires_grad)
+    # At the end so that the conj() is tracked by the autograd
```
Why do we want `conj()` tracked by autograd?

Lezcano replied:
because it needs to be tracked by autograd for everything to work as expected I believe. At least, that's what's done in the OpInfos that generate conjugate inputs. I could try deactivating the tracking of the autograd in both cases if that's any better. Any thoughts @anjali411?

anjali411 replied:
I don't think it's necessary for `conj` to be tracked by autograd for these tests. We have an `OpInfo` for conj to verify it behaves correctly independently. We do want to run gradcheck with conjugated inputs, but that's about it.

mruberry commented on test/test_ops.py:
```diff
@@ -244,34 +244,47 @@ def test_noncontiguous_samples(self, device, dtype, op):
+                n_inputs = (n_inp,) + n_args if isinstance(n_inp, torch.Tensor) else tuple(n_inp) + n_args
+                for j, (t, n) in enumerate(zip(t_inputs, n_inputs)):
+                    if isinstance(t, torch.Tensor) and t.requires_grad:
+                        msg = ("{}Got different gradients for contiguous / non-contiguous wrt input {}."
```
Won't this format like:

"1Got different ..."?

Lezcano replied:
Note that I'm adding a space `"Output {i}. "`

mruberry commented on test/test_ops.py:
```diff
@@ -244,34 +244,47 @@ def test_noncontiguous_samples(self, device, dtype, op):
+                    inp = [inp]
+                for t in inp:
+                    if isinstance(t, torch.Tensor) and t.requires_grad and t.grad is not None:
+                        t.grad.zero_()
```
Setting the grad to None might be faster

Lezcano replied:
I thought about this, I don't knwo why didn't I do it this way. I'll try this next week, see if it goes through or there was a reason for why I didn't use `None`.

mruberry commented on test/test_ops.py:
```diff
@@ -244,34 +244,47 @@ def test_noncontiguous_samples(self, device, dtype, op):
-            elif isinstance(expected, Sequence) and isinstance(expected[0], torch.Tensor):
-                expected_backward_tensor = expected[0]
-                actual_backward_tensor = actual[0]
+                assertEqualGradients(expected, actual)
```
Nice generalization

mruberry commented on test/test_ops.py:
```diff
@@ -244,34 +244,47 @@ def test_noncontiguous_samples(self, device, dtype, op):
             if not test_grad:
                 continue
 
+            def assertEqualGradients(expected, actual, i=None):
```
Did you walk through an example locally to ensure this is doing what you expect? 

These tests are so complicated I always have to do that to feel comfortable they're testing anything

Lezcano replied:
Yep, and I did not find anything weird.

They also failed when the generation of the inputs was not done in an independent manner (i.e. some inputs were views of others) which is reassuring.

Lezcano replied:
I have heavily simplified this test, and now it should be easier to reason about it. Could you have another look?

mruberry replied:
The updated design looks great!

mruberry reviewed (approved):
Really nice generalization

mruberry commented:
I'm a little surprised this test extension didn't uncover any new issues, or does the previous PR updating the sample inputs for the indexing functions preemptively resolve issues this would have hit?

Lezcano commented:
In this and the previous PRs, I just cleaned the generation of inputs in the OpInfos. I was also a bit surprised that this did not uncover any problems. In the case of `linalg`, this is expected, as many operations were tested with noncontiguous inputs (see the last PR in this stack that removes those tests). In other operations... I guess that you already uncovered all the problematic operations with your tests? I do not know...

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

anjali411 commented on torch/testing/_internal/common_utils.py:
```diff
@@ -2423,27 +2423,36 @@ def noncontiguous_like(t):
+
+    # Handle conj bit. repeat_interleave resolves the conj, which is a bit annoying.
+    is_conj = t.is_conj()
+    if is_conj:
```
should add a similar change for `neg` bit too, no?

anjali411 reviewed (commented):
Let's update `repeat_interleave` to not materialize conjugation or negation (see my comment above)

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

Lezcano commented:
To address the reviews, I prepended 2 PRs to this stack:
https://github.com/pytorch/pytorch/pull/68522
https://github.com/pytorch/pytorch/pull/68523
Could you (@anjali411 @mruberry) have a look at them?

I also updated this PR accordingly, as the logic around `repeat_interleave` is no longer necessary.

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

mruberry reviewed (approved):
Cool!

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

[ghstack-poisoned]

Lezcano commented:
Rebased @mruberry

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33774418).

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

Differential Revision: [D33774418](https://our.internmc.facebook.com/intern/diff/D33774418)

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33774418).

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

Differential Revision: [D33774418](https://our.internmc.facebook.com/intern/diff/D33774418)

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33774418).

lezcano committed:
Update on "Generalize noncontiguous tests to several outputs"


This is necessary for most matrix decompositions in `linalg`.

cc mruberry

Differential Revision: [D33774418](https://our.internmc.facebook.com/intern/diff/D33774418)

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33774418).
